src/modules/commander/state_machine_helper.cpp

```diff
@@ -985,6 +993,43 @@ void reset_link_loss_globals(actuator_armed_s &armed, const bool old_failsafe, c
 	}
 }
 
+void set_quadchute_nav_state(vehicle_status_s &status, actuator_armed_s &armed,
+			     const vehicle_status_flags_s &status_flags,
+			     const quadchute_actions_t quadchute_act)
+{
+	switch (quadchute_act) {
+
+	// add DISABLED, default RTL
+	case quadchute_actions_t::DISABLED:
+		// If quadchute action is disabled then no action must be taken.
+		return;
+
+	case quadchute_actions_t::AUTO_RTL:
+		if (status_flags.condition_global_position_valid && status_flags.condition_home_position_valid) {
+			status.nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_RTL;
+			return;
+		}
+
+	// FALLTHROUGH
+	case quadchute_actions_t::AUTO_LAND:
+		if (status_flags.condition_global_position_valid) {
+			status.nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LAND;
+			return;
+		}
+
+	// FALLTHROUGH
+	case quadchute_actions_t::AUTO_LOITER:
+		if (status_flags.condition_global_position_valid) {
+			status.nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LOITER;
+			return;
+		}
+	}
+
+	// If none of the above worked, force to AUTO_RTL mode
```

Comment:
If you want I can also remove these lines such that if the desired action is not possible, no actions are taken rather than falling back to the default action.

Reply:
> Does this "enforcing" of a mode also happen in other places?

Ah wait, I only saw what enforce means that after a second read. You could set the nav state RTL even if position is not valid. That's a pretty bad idea and probably leads to a hole where things go really wrong.

> what's best practice in your opinion?

Thanks for asking.

- You probably want to catch a misconfiguration. If something has to be done for safety even if the parameter is set to a value that is out of range do the default action. I would have just written it a bit differently to cover the default case of the switch and not use any early returns.
- With the fallbacks we have to be super careful because if we set the nav state to RTL with no valid position or home commander failed its job. Sadly it's not super consistent yet but check_invalid_pos_nav_state() covers the position checks part all the way falling down to terminate. And you check correctly for the home position. Falling back from land to loiter doesn't make sense because the fallback can never happen (same conditions). It should rather do no action if none of it is possible. I can't think of a case where the fallbacks of the original mode wouldn't do the right thing if the action is not possible.

I can make a suggestion that then @JonasPerolini can decide to use or not.

Reply:
Thank you for your reply @MaEtUgR . I agree that if the position is not valid, it's dangerous to try an RTL. I have thus removed the enforcement of an RTL. I've also modified the early returns and covered the default case of the switch as you proposed. This way, if we have a misconfiguration, we will try to change the nav state to RTL and if it's not possible (i.e. pose not valid) we will not take any action.

Reply:
Nice, thanks!